lib/AST/ASTWalker.cpp

```diff
@@ -1341,6 +1344,14 @@ class Traversal : public ASTVisitor<Traversal, Expr*, Stmt*,
     std::tie(shouldWalkArguments, shouldWalkExpansion) =
         Walker.shouldWalkMacroArgumentsAndExpansion();
 
+    if (auto *substituteDecl = E->getSubstituteDecl()) {
+      if (doIt(substituteDecl))
+        return nullptr;
+      // Visiting the substitute macro expansion decl will visit the same
+      // argument list. Skip visiting it again.
+      shouldWalkArguments = false;
+    }
+
```

Comment:
@DougGregor Because an `MacroExpansionExpr` and its substitute decl are sharing the argument list again after [`1ba7c14`](https://github.com/apple/swift/pull/64761/files#diff-1ba7c144bdfaf2daf57703f890ef9575a5782850e77c573df4170a547d73ab99L5402), this became necessary to prevent the argument list from being visited twice. Otherwise it triggers an assertion failure in `VarDeclUsageChecker`.

Reply:
This localized suppression of visiting the argument list makes sense to me, thanks for pointing it out.